lazy_dataset/parallel_utils.py

```diff
@@ -201,11 +203,27 @@ def result(job: concurrent.futures.Future):
             return job.result()
 
         def terminate(ex: concurrent.futures.Executor, q):
-            # shutdown doesn't work for threads. Don't know why.
-            # For mp shutdown doesn't work and the processes keep the
-            # program alive, i.e. it never stops properly.
-            # Hence, use cancel for both.
+            # From https://docs.python.org/3/library/concurrent.futures.html
+            #
+            # > Regardless of the value of wait, the entire Python program
+            # > will not exit until all pending futures are done executing.
+            #
+            # This is an undesired behaviour for lazy_dataset, i.e., when
+            # the garbage collector wants to delete the executor, the
+            # executor will finish first all calculations.
+            # Hence, cancel all futures, so it will at least not start new
+            # "tasks".
+            #
+            # For threads shutdown is intended to not terminate submitted
+            # tasks.
+            # For mp shutdown also doesn't work as desired and the processes
+            # keep the program alive, i.e. it never stops properly.
+            # Hence, use only cancel for both.
             # ex.shutdown(wait=False)
```

Comment:
Not sure: Should `shutdown` still be called? Even if it doesn't cancel the tasks it prevents new tasks from being submitted to the pool. I feel like it is not required since we have full control over the pool and won't submit new tasks after calling `terminate`, but it's the cleanest way to signal termination to the threadpool

Reply:
Now that I looked at the code I think `shutdown` should be called. It makes sure that all threads wake (https://github.com/python/cpython/blame/365e0772c9d9be9bd70379afa0bf01ecb9bbe744/Lib/concurrent/futures/thread.py#L229) and no thread keeps waiting on the (empty) queue.